pkg/operator/staticpod/controller/installerstate/installer_state_controller.go

```diff
@@ -0,0 +1,254 @@
+func (c *InstallerStateController) eventHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.queue.Add(installerStateControllerWorkQueueKey) },
+		UpdateFunc: func(old, new interface{}) { c.queue.Add(installerStateControllerWorkQueueKey) },
+		DeleteFunc: func(obj interface{}) { c.queue.Add(installerStateControllerWorkQueueKey) },
+	}
+}
+
+// degradedConditionNames lists all supported condition types.
+var degradedConditionNames = []string{
+	"InstallerPodPendingDegraded",
+	"InstallerPodContainerWaitingDegraded",
+	"InstallerPodNetworkingDegraded",
+}
+
+func (c *InstallerStateController) sync() error {
+	pods, err := c.podsGetter.Pods(c.targetNamespace).List(metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set{"app": "installer"}).String(),
+	})
+	if err != nil {
+		return err
+	}
+
+	// collect all startingObjects that are in pending state for longer than maxToleratedPodPendingDuration
+	pendingPods := []*v1.Pod{}
+	for _, pod := range pods.Items {
+		if pod.Status.Phase != v1.PodPending || pod.Status.StartTime == nil {
+			continue
+		}
+		if c.timeNowFn().Sub(pod.Status.StartTime.Time) >= maxToleratedPodPendingDuration {
+			pendingPods = append(pendingPods, pod.DeepCopy())
+		}
+	}
+
+	// in theory, there should never be two installer startingObjects pending as we don't roll new installer pod
+	// until the previous/existing pod has finished its job.
+	foundConditions := []operatorv1.OperatorCondition{}
+	foundConditions = append(foundConditions, c.handlePendingInstallerPods(pendingPods)...)
+
+	// handle networking conditions that are based on events
+	networkConditions, err := c.handlePendingInstallerPodsNetworkEvents(pendingPods)
+	if err != nil {
+		return err
+	}
+	foundConditions = append(foundConditions, networkConditions...)
+
+	updateConditionFuncs := []v1helpers.UpdateStaticPodStatusFunc{}
+
+	// check the supported degraded foundConditions and check if any pending pod matching them.
+	for _, degradedConditionName := range degradedConditionNames {
+		// clean up existing foundConditions
+		updatedCondition := operatorv1.OperatorCondition{
+			Type:   degradedConditionName,
+			Status: operatorv1.ConditionFalse,
+		}
+		if condition := v1helpers.FindOperatorCondition(foundConditions, degradedConditionName); condition != nil {
+			updatedCondition = *condition
+		}
+		updateConditionFuncs = append(updateConditionFuncs, v1helpers.UpdateStaticPodConditionFn(updatedCondition))
+	}
+
+	if _, _, err := v1helpers.UpdateStaticPodStatus(c.operatorClient, updateConditionFuncs...); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *InstallerStateController) handlePendingInstallerPodsNetworkEvents(pods []*v1.Pod) ([]operatorv1.OperatorCondition, error) {
+	conditions := []operatorv1.OperatorCondition{}
+	if len(pods) == 0 {
+		return conditions, nil
+	}
+	namespaceEvents, err := c.eventsGetter.Events(c.targetNamespace).List(metav1.ListOptions{})
```

Comment:
Wouldn't it be better to use `Search(scheme, obj)` here? This way you'd get only events for the pods. You would be exchanging one call vs several but each will be about specific object?
I'm not sure if obj can be a list of objects, though.

Reply:
i'm not sure the .Search will make it faster/more efficient... this is only listing events for single namespace and only when there is actual pending pods :-)